launch_ros/launch_ros/ros_adapters.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2018-2020 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module containing ROS specific adapters and their initialization."""
+
+import os
+import threading
+from typing import List
+from typing import Optional
+
+import launch
+import launch.events
+
+import rclpy
+from rclpy.executors import SingleThreadedExecutor
+
+
+class ROSAdapter:
+    """Wraps rclpy API to ease ROS node usage in `launch_ros` actions."""
+
+    def __init__(
+        self,
+        *,
+        argv: Optional[List[str]] = None,
+        context: Optional[rclpy.Context] = None,
+        autostart: bool = True
+    ):
+        """
+        Construct adapter.
+
+        :param: argv List of global arguments for rclpy context initialization.
+        :param: context Provide a context other than the default rclpy context
+          to pass down to rclpy.init.
+          The context is expected to have already been initialized by the caller
+          using rclpy.init().
+        :param: autostart Whether to start adapter on construction or not.
+        """
+        self.__argv = argv
+        self.__context = context
+        self.__node = None
+        self.__executor = None
+        self.__is_running = False
+
+        if autostart:
+            self.start()
+
+    def start(self):
+        """Start ROS adapter."""
+        if self.__is_running:
+            raise RuntimeError('Cannot start a ROS adapter that is already running')
+        try:
+            if self.__context is not None:
```

Comment:
I think that if context is `None`, then this action should create and manager its own context, rather than passing the `None` on to `create_node` so that the default singleton is used.

Imagine:

- user does rclpy.init()
- this function trys that but catches the error that it is already initialized
- this function creates node with `context=None` (use singleton context)
- user does rclpy.shutdown()
- this class or an action using this class trys to do something with the shutdown context

By managing its own context, nothing the user does will affect it.

Reply:
That makes sense, and goes hand-in-hand with `rclpy` contexts used by `launch` should have their lifetime handled by `launch`.

Reply:
Thinking about this again, do we even need to allow for a context to be passed along? Without an `ExplicitROSStartup` action, users can't access this class' instance. Arguments can already be provided to initialize it and context is accessible through a property. So if you don't disagree, I'll just drop that.

Reply:
Done in 577d77bfc5fa2101616ce7856b6d55745370692b.

Reply:
I don't think we need to, but having the ability to doesn't cost anything, imo.

Reply:
Rationale for dropping it was that it'd require additional logic to decide what should handle that context's lifetime, and for no reason since we've agreed these adapters are for actions to build on and not for users to play with.

Reply:
That's fine with me to leave it as-is.